src/ts/utils/dom.ts

```diff
@@ -1,41 +1,116 @@
+    },
+    
+    
+    getInputEvent() {
+        return new Event('input', {
+            bubbles: true,
+            cancelable: true,
+        });
+    },
+    
+    async detectChange(fn: () => void, el?: HTMLElement){
+        const targetNode = el || document.querySelector('.roam-body-main') as HTMLElement;
+        const config = { attributes: true, childList: true, subtree: true };
+        return new Promise(async resolve => {
+            //@ts-ignore unused arg
+            const callback = function(mutationsList:MutationRecord[], observer: MutationObserver) {
+                resolve('mutated')
+                observer.disconnect();
+            };
+            const observer = new MutationObserver(callback);
+            observer.observe(targetNode, config);
+            await fn();
+        })
+    },
+    
+    getBlockContainer(element?: HTMLElement) {
+        let parent = null; 
+        let el = element || document.querySelector('textarea') as HTMLElement
+        while(el && !parent){
+            if (el.parentElement?.className?.includes('flex-v-box roam-block-container')) {
+                parent = el?.parentElement;
+            }
+            el = el?.parentElement as HTMLElement;
+        }
+        return parent as HTMLElement
+    },
+    
+    getBlockParent(element: HTMLElement){
+        return this.getBlockContainer(this.getBlockContainer(element));
+    },
+    
+    isTopLevelBlock(element: HTMLElement){
+        return !this.getBlockParent(element);
+    },
+    
+    hasChildren(element: HTMLElement){
+        return !!this.getBlockChildren(element)?.length;
+    },
+    
+    getBlockChildren(element: HTMLElement) : HTMLCollection{
+        const childrenDiv = this.getBlockContainer(element)?.lastChild as HTMLElement;
+        return childrenDiv?.children;
+    },
+    getNthChild(element: HTMLElement, n:number){
+        return this.getBlockChildren(element)?.[n].querySelector('.roam-block') as HTMLElement;
+    },
+    getFirstChild(element: HTMLElement){
+        return this.getNthChild(element, 0);
+    },
+    getLastChild(element: HTMLElement){
+        return this.getNthChild(element,this.getBlockChildren(element).length-1);
+    },
+    hasSiblings(element: HTMLElement) {
+        return !!this.getSiblings(element)
+    },
+    getSiblings(element: HTMLElement):  HTMLCollection {
+        if (this.isTopLevelBlock(element)) {
+            return this.getBlockContainer(element)?.parentElement?.children as HTMLCollection
+        } else {
+            return this.getBlockChildren(this.getBlockParent(element))
+        }
+    },
+    getNthSibling(element:HTMLElement, n:number) {
+        return this.getSiblings(element)?.[n].querySelector('.roam-block');
+    },
+    getFirstSibling(element: HTMLElement){
+        return this.getNthSibling(element, 0);
+    },
+    getLastSibling(element: HTMLElement){
+        return this.getNthSibling(element,this.getSiblings(element).length-1);
+    },
 }
```

Comment:
I think we should just apply `.querySelector('.roam-block');` before we return results from `getSiblings` and return the array of that. Then we won't need these specialized methods.

Similarly for the children